### DISCUSSION TITLE
Lock file interop between npm 6 and 7

### DISCUSSION THREAD

**Discussion by targos** [npm/feedback#59] [CLI] [answered]
To ease the transition of people from npm 6 to npm 7, I think it would be really useful that npm 6 knows a little bit more about the new lock file version.

From what I have experienced, v2 introduces:
- A new "packages" field: npm 6 could just ignore it
- A new sorting algorithm for the "dependencies" fields: maybe npm 6 could support that too if `lockfileVersion` is equal to 2.

What I would like is that when npm 6 is used to install dependencies with a lock file created with npm 7, it doesn't completely change the contents of the lock file. This is currently preventing me from testing npm 7 extensively in my projects.

**Comment by MylesBorins** [comment]
pinging the @npm/cli team about this.

**Comment by isaacs** [comment]
Yeah, the design of `lockfileVersion: 2` is intended to not _break_ npm v6, and I think it achieves that.  The different sorting algorithm is easy enough to do, it's basically just using `String.localeCompare` rather than whatever npm v6 is doing.

One relatively easy thing would be for npm v6 to remove the `packages` field, only update the legacy metadata in `dependencies` (as it does today), use the same sorting, and not complain about it.  The much harder thing would be to update npm v6 to also update the `packages` field appropriately, since that object tracks much more data about the packages on disk.  The worst thing we could do is have npm v6 update `dependencies` but not keep `packages` in sync, since then npm v7 and v6 will be installing different trees.

The easiest thing is to not update it at all, of course, since that's zero work.  The slight advantage there would be that it provides some incentive for everyone on the team to update to npm v7, and this is the surest way to avoid situations where the lockfile metadata is incomplete or the tree is less correctly designed.  Since npm v6 is a temporary problem (albeit a frustratingly long-lived one), eventually it'll stop being an issue, but our intention is definitely to make it as easy as possible for teams to update, without requiring that they do so all at the same exact time.

**Reply by MylesBorins** [reply]
@targos do you have any additional questions?

**Comment by MylesBorins** [comment, marked as answer]
`npm_config_save=false` as an env var or `--no-save` or `--save=false` can all be used to avoid updating the lock file when installing with npm 6 or 7. For now this seems to be the best solution to avoid lots of churn. Anything else would require quite a bit of work for npm 6.
